rust/pact_consumer/src/mock_server.rs

```diff
@@ -178,6 +140,148 @@ impl ValidatingMockServer {
+    /// pact, and runs in a background thread
+    pub fn new(pact: Pact) -> BackgroundMockServer {
+        let mut runtime = tokio::runtime::Builder::new()
+            .threaded_scheduler()
+            .core_threads(1)
+            .enable_all()
+            .build()
+            .unwrap();
+
+        let (mock_server, future) = runtime.block_on(InnerServer::init_mock_server(pact))
+            .expect("error initializing mock server");
+
+        BackgroundMockServer {
+            inner: InnerServer::new(mock_server, runtime.spawn(future)),
+            runtime,
+        }
+    }
+
+    fn drop_helper(&mut self) -> Result<(), String> {
+        let server_handle = self.inner.server_handle
+            .take()
+            .expect("Server was already shut down");
+        self.inner.mock_server.shutdown().ok();
+
+        self.runtime.block_on(server_handle).unwrap();
+
+        self.inner.drop_helper()
+    }
+}
+
+impl ValidatingMockServer for BackgroundMockServer {
+    fn url(&self) -> &Url {
+        &self.inner.url
+    }
+
+    fn path<P: AsRef<str>>(&self, path: P) -> Url {
+        // We panic here because this a _test_ library, the `?` operator is
+        // useless in tests, and filling up our test code with piles of `unwrap`
+        // calls is ugly.
+        self.inner.url.join(path.as_ref()).expect("could not parse URL")
+    }
+
+    fn status(&self) -> Vec<MatchResult> {
+        self.inner.mock_server.mismatches()
+    }
+}
+
+impl Drop for BackgroundMockServer {
+    fn drop(&mut self) {
+        let result = self.drop_helper();
+        if let Err(msg) = result {
+            panic_or_print_error(&msg);
+        }
+    }
+}
+
+///
+/// A mock server implementation that is spawned onto the implicit current
+/// tokio runtime that drives the current unit test via tokio::test.
+/// After use, the server must be explitcly and asynchronously disposed by a call to
+/// shutdown(). If not, the test runtime might end up deadlocked.
+///
+#[must_use = "Remember to call shutdown().await"]
+pub struct SpawnedMockServer {
+    inner: InnerServer
+}
+
+impl SpawnedMockServer {
+    /// Asynchronously create and spawn a new mock server
+    pub async fn new(pact: Pact) -> SpawnedMockServer {
+        let (mock_server, future) = InnerServer::init_mock_server(pact)
+            .await
+            .expect("error initializing mock server");
+
+        SpawnedMockServer {
+            inner: InnerServer::new(mock_server, tokio::spawn(future))
+        }
+    }
+
+    /// Asynchronously shut down the server.
+    /// This is necessary before the test runtime gets destroyed.
+    /// The reason is that the mock server is spawned as a background task.
+    /// When the tokio test runtime tries to shut down, it will wait for
+    /// all its running tasks to complete. Only by first awaiting the server's
+    /// JoinHandle can the shutdown procedure be properly executed.
+    pub async fn shutdown(mut self) {
```

Comment:
It appears that this design is flawed after all. While the server shuts down cleanly on test success, it panics when a test panics. That's a panic while panicing, a very bad idea. I think I'll remove the async functionality again, instead focusing on how to reliably run the server in a separate thread. That's the strategy of `reqwest`s own integration tests, so I'll try to copy that behaviour.